src/components/uart/hardware.cpp

```
@@ -0,0 +1,214 @@
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_8E2:
+    return SERIAL_8E2;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_5O1:
+    return SERIAL_5O1;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_6O1:
+    return SERIAL_6O1;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_7O1:
+    return SERIAL_7O1;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_8O1:
+    return SERIAL_8O1;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_5O2:
+    return SERIAL_5O2;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_6O2:
+    return SERIAL_6O2;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_7O2:
+    return SERIAL_7O2;
+  case wippersnapper_uart_UartPacketFormat_UART_PACKET_FORMAT_8O2:
+    return SERIAL_8O2;
+  default:
+    return SERIAL_8N1;
+  }
+}
+
+/*!
+ * @brief  Configures the UART serial using a provided configuration.
+ * @return True if the serial was successfully configured, False otherwise.
+ */
+bool UARTHardware::ConfigureSerial() {
+  int8_t rx_pin = -1;
+  int8_t tx_pin = -1;
+  if (_config.pin_rx[0] == 'D' || _config.pin_rx[0] == 'A') {
+    rx_pin = atoi(_config.pin_rx + 1);
+  } else {
+    rx_pin = atoi(_config.pin_rx);
+  }
+  if (_config.pin_tx[0] == 'D' || _config.pin_tx[0] == 'A') {
+    tx_pin = atoi(_config.pin_tx + 1);
+  } else {
+    tx_pin = atoi(_config.pin_tx);
+  }
+  uint16_t cfg = UartPacketFormatToConfig(_config.format);
+
+  if (_config.use_sw_serial) {
+// Does this architecture support SoftwareSerial?
+#if HAS_SW_SERIAL
+    // Create a new SoftwareSerial instance
+    _swSerial = new SoftwareSerial(rx_pin, tx_pin, _config.sw_serial_invert);
+    if (_swSerial == nullptr) {
+      WS_DEBUG_PRINTLN(
+          "[uart] ERROR: Failed to allocate a SoftwareSerial instance!");
+      return false;
+    }
+    _swSerial->begin((unsigned long)_config.baud_rate);
+    _baud_rate = _config.baud_rate;
+#endif // HAS_SW_SERIAL
```

Should we raise an error here if software serial is requested but unavailable on this platform?

---

No because HAS_SW_SERIAL is defined by the boards.h file

---

Brent Rubell
[@Tyeth]( ) I’m confused by [this comment](https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/787#discussion_r2261244145). rp2040 has hardware UART


Tyeth
This bit above (sorry for bad link) says if rp2040 add softserial [https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/787/files#diff-7eff18bcfa8b3bbf92bd0830bfab7c[…]a75f37a5a645fec88746aa4daa9b5R24-R36](https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/787/files#diff-7eff18bcfa8b3bbf92bd0830bfab7c76d6ca75f37a5a645fec88746aa4daa9b5R24-R36)
Is that just future use and we use HW serial for uart on rp2040? (edited) 


Brent Rubell
no it uses the swserial wrapper on rp2040
for uart


Tyeth
So does that mean using software serial for UART gps and pm25 and distance sensor components? (on rp2040)


Brent Rubell
for uart components (non gps yes).
gps doesn’t have swserial impl’d at the moment.
I think I can add it, let me use the bathroom and timebox this to 45min. It shouldn’t be difficult because ws_uart already exposes the swserial wrapper.

---

Implemented SoftwareSerial support within UART drivers in 5ef64390..6a54d02b